### PR TITLE
feat: enable bean discovery when bean-discovery-mode="annotated"

### LINKED ISSUE #303: @RouteScoped beans are not discovered when bean-discovery-mode="annotated"
When using bean-discovery-mode="annotated" all beans with Pseudo-scopes are not discovered or considered as beans. In order to mitigate this RouteScoped would have to be a Stereotype for example. This is e.g. discussed in https://issues.jboss.org/browse/WELD-1733?_sscc=t

```java
@Stereotype
@Scope
@Inherited
@Target({ ElementType.ANNOTATION_TYPE, ElementType.TYPE, ElementType.FIELD,
        ElementType.METHOD, ElementType.CONSTRUCTOR })
@Retention(RetentionPolicy.RUNTIME)
public @interface RouteScoped {
}
```

### LINKED ISSUE #391: Constructor Injection only works with CDI Beans
According to the documenation at https://vaadin.com/docs/latest/integrations/cdi/instantiated-beans
> If the type is not found as a CDI bean – for example, because it is ambiguous or does not have a no-arguments public constructor – instantiation falls back to the default Vaadin behavior, and the component is instantiated as a plain-old Java object (POJO). Dependency injection is performed after instantiation, and injects still work, but other CDI features do not. The reason for this is that the instantiated component is not a contextual instance.

Actually, if the type is not found as a CDI bean
- a no-args or var-args constructor is required, see [ReflectTools.createInstance(Class)](https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/internal/ReflectTools.java#L480)
- only field injection is performed, see [AbstractCdiInstantiator.getOrCreate(Class)](https://github.com/vaadin/cdi/blob/master/vaadin-cdi/src/main/java/com/vaadin/cdi/AbstractCdiInstantiator.java#L67)

Consequently

- constructor injection only works with CDI beans
- the flow-cdi-tutorial and the examples on https://vaadin.com/docs/latest/integrations/cdi/contexts only work with [bean-discovery-mode="all"](https://github.com/vaadin/flow-cdi-tutorial/blob/master/src/main/webapp/WEB-INF/beans.xml) because they use constructor injection

Please update the documentation or treat it as a bug.

### PR DESCRIPTION
**Pull request by taefi**
## Description
@CdiComponent added to be used on types
that are supposed to be discovered and
managed by CDI container when
bean-discovery-mode is set to "annotated"
in beans.xml

Fixes #303
Fixes #391

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

**Comment by roeltje25** [comment]
Thanks for providing a solution to #303 this way., although I would have argued fervently for #392. But reading the specs I may interpret the intentions for @Stereotype a bit broader than you guys do. Fair enough.  Thanks for picking this up!

**Comment by roeltje25** [comment]
Ok, one more little argument for why #392 is actually a good solution and is not a case of not-intended use of @Stereotype. Please look at the implementation of e.g. the @Decorator stereotype in package javax.decorator. This is a stereotype, in order to be found by CDI, but it does not declare any additional scope, @Alternative, or @Named annotation. Clearly, the  meaning of the annotation is to make @Decorator a bean defining annotation. 
Also a @Stereotype can according to its javadoc have 0 interceptor bindings, 0 scope annotations, 0 alternative and 0 named and 0 stereotypes.

Ok. anyway. I will stop.

File: vaadin-cdi-itest/src/test/java/com/vaadin/cdi/itest/BeanDiscoveryModeTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    @OperateOnDeployment("bean-discovery-mode-all")
+    public void beanDiscoveryWorks_when_beanDiscoveryModeIsAll() {
+        validateBeanDiscoveryAndInjectionWorks();
+    }
+
+    private void validateBeanDiscoveryAndInjectionWorks() {
+        click(SET_NAME_BTN_ID);
+        click(NORMAL_SRV_BTN_ID);
+        waitUntilNot(driver -> getText(RESULT_SPAN_ID) == null);
+        Assert.assertEquals("Hello MyName", getText(RESULT_SPAN_ID));
```
Comment (mcollovati) [inline vaadin-cdi-itest/src/test/java/com/vaadin/cdi/itest/BeanDiscoveryModeTest.java:69]: What about making the two service returns a different output, just to be sure that the text of `RESULT_SPAN_ID` has been effectively changed?
For example, instead of just `Hello xyz`, return `Hello from (CDIComponent|NormalScoped), xyz`

File: vaadin-cdi/src/main/java/com/vaadin/cdi/annotation/CdiComponent.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import java.lang.annotation.*;
+
+/**
+ * This is a {@code @Stereotype} to let the container scan and manage
+ * the instances of the annotated types.
+ * <p>
+ * Annotating the beans with this annotation is necessary when the
+ * {@code bean-discovery-mode="annotated"} is set in the beans.xml
+ * file (implicit bean archive), and in this case CDI can only manage
+ * and inject beans annotated with a "scope" type.
```
Comment (mcollovati) [inline vaadin-cdi/src/main/java/com/vaadin/cdi/annotation/CdiComponent.java:28]: > can only manage and inject beans annotated with a "scope" type.

I cannot understand this statement. Do you mean beans annotated with a "normal scope"? In this case, I would rephrase as `can only manage and inject beans annotated with a bean defining annotations`.

Comment (taefi) [reply vaadin-cdi/src/main/java/com/vaadin/cdi/annotation/CdiComponent.java]: That statement is coming from https://docs.oracle.com/javaee/7/tutorial/cdi-adv001.htm. But, I agree that rephrasing it to what you proposed can make it clearer.

**Review by mcollovati — APPROVED** [review]
